### PR TITLE
Use builder directly for OpIAddCarry, OpISubBorrow, Op[SU]MulExtended

### PR DESCRIPTION
**Pull request by trenouf**
This commit changes the SPIR-V reader to generate IR directly for the
above operations.

Change-Id: Ia2ac290bc4f21ca0497f85ac51a3dd0d146eee9e

### CODE REVIEW

**Comment by trenouf** [comment]
Please review just the topmost commit.

File: translator/lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -6008,6 +6008,66 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
...
+                               Val0->getType()),
+        1);
+    return mapValue(BC, Result);
+  }
+
+  case OpISubBorrow: {
+    SPIRVBinary *BC = static_cast<SPIRVBinary *>(BV);
+    Value *Val0 = transValue(BC->getOperand(0), F, BB);
+    Value *Val1 = transValue(BC->getOperand(1), F, BB);
+    Value *Add = getBuilder()->CreateIntrinsic(Intrinsic::usub_with_overflow,
```
Comment (amdrexu) [inline translator/lib/SPIRV/SPIRVReader.cpp]: Add -> Sub

File: translator/lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -6008,6 +6008,66 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
     return mapValue(BC, Result);
   }
 
+  case OpIAddCarry: {
+    SPIRVBinary *BC = static_cast<SPIRVBinary *>(BV);
+    Value *Val0 = transValue(BC->getOperand(0), F, BB);
+    Value *Val1 = transValue(BC->getOperand(1), F, BB);
+    Value *Add = getBuilder()->CreateIntrinsic(Intrinsic::uadd_with_overflow,
+                                             Val0->getType(), {Val0, Val1});
```
Comment (amdrexu) [inline translator/lib/SPIRV/SPIRVReader.cpp]: {Val0, Val1} ->  { Val0, Val1 }

File: translator/lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -6008,6 +6008,66 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
...
+        1);
+    return mapValue(BC, Result);
+  }
+
+  case OpISubBorrow: {
+    SPIRVBinary *BC = static_cast<SPIRVBinary *>(BV);
+    Value *Val0 = transValue(BC->getOperand(0), F, BB);
+    Value *Val1 = transValue(BC->getOperand(1), F, BB);
+    Value *Add = getBuilder()->CreateIntrinsic(Intrinsic::usub_with_overflow,
+                                             Val0->getType(), {Val0, Val1});
```
Comment (amdrexu) [inline translator/lib/SPIRV/SPIRVReader.cpp]: {Val0, Val1} ->  { Val0, Val1 }

File: translator/lib/SPIRV/SPIRVReader.cpp (hunk 2 above)
Comment (trenouf) [reply translator/lib/SPIRV/SPIRVReader.cpp]: That's wrong for llvm style. Running clang-format on this code leaves no whitespace after the { or before the }.

**Comment by trenouf** [comment]
V2: Rebase
V3: Addressed review comments.

File: translator/lib/SPIRV/SPIRVReader.cpp (hunk 2 above)
Comment (amdrexu) [reply translator/lib/SPIRV/SPIRVReader.cpp]: I think we ought to make our own version with some modifications of clang-format. In driver development, place spaces after { or before } is to make the array data clear. Also, there are other places that I'm unhappy with default clang-format: 80-character line limitation, 2 spaces indentation, acronym without camelcase (leading to this ABCDEfgh, not AbcdEfgh)

**Comment by amdvlk-admin** [comment]
### Test summary for commit 80618aa4a01dc8cb77ac71dbf41f381c5d7f35eb
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev dcb42575c3d374467bb6e4e609f8bf6d1302c297</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**